SE-284 [LX-71] Add CI build to HTML XBlock

ArtemBernatskyy opened the pull request:
**JIRA tickets**: https://tasks.opencraft.com/browse/SE-325 , https://tasks.opencraft.com/browse/LX-71

**Author notes and concerns**:  don't know how to test `xblocks` as for now so created a dummy test
**Reviewers**: @symbolist

symbolist commented on LICENSE:
```diff
@@ -1,671 +0,0 @@
-                    GNU AFFERO GENERAL PUBLIC LICENSE
```
We will be going with AGPL so you can leave this in.

symbolist commented on README.md:
```diff
@@ -1,5 +1,8 @@
 # HTML XBlock
 
+[![build](https://circleci.com/gh/ArtemBernatskyy/xblock-html/tree/wozniak%2Fci-setup.svg?style=shield)](https://circleci.com/gh/ArtemBernatskyy/xblock-html/tree/wozniak%2Fci-setup) [![codecov](https://codecov.io/gh/ArtemBernatskyy/xblock-html/branch/wozniak%2Fci-setup/graph/badge.svg)](https://codecov.io/gh/ArtemBernatskyy/xblock-html)
```
Can you update these to point to the opencraft org? I have enabled CircleCI for this repo.

symbolist commented on requirements/base.txt:
```diff
@@ -0,0 +1,8 @@
+git+https://github.com/edx/i18n-tools.git
+django-statici18n==1.8.0
+edx-i18n-tools==0.4.7
+Mako==1.0.7
```
This is not needed so you can remove it.

ArtemBernatskyy replied:
We need this to prevent CI from crashing

symbolist commented on requirements/base.txt:
```diff
@@ -0,0 +1,8 @@
+# Requirements for app run
+
+git+https://github.com/edx/xblock-utils.git@v1.1.1#egg=xblock-utils==1.1.1
+git+https://github.com/edx/i18n-tools.git
```
Please pin this also.

symbolist commented on setup.py:
```diff
@@ -0,0 +1,40 @@
+    name='xblock-html',
+    version='0.1',
+    description='xblock-html',   # TODO: write a better description.
+    license='UNKNOWN',          # TODO: choose a license: 'AGPL v3' and 'Apache 2.0' are popular.
```
You can set this to AGPL v3.

symbolist commented on tests/test_temp.py:
```diff
@@ -0,0 +1,10 @@
+"""
+Just dummy test
```
You can call this file `tests.py` and then the tests can be added to it later.

symbolist commented on .circleci/config.yml:
```diff
@@ -0,0 +1,57 @@
+      - run:
+          name: Installing requirements
+          command: |
+            virtualenv env
```
Can you move the env out of the source dir? See https://github.com/edx-olive-oc/edx-platform/commit/e5a883e3c5024cc4a5c52a7b9b94d7bb44044589#diff-1d37e48f9ceff6d8030570cd36286a61

symbolist commented on .circleci/config.yml:
```diff
@@ -0,0 +1,57 @@
+          name: Running tests
+          command: tox
+      - run:
+          name: Test Quality
```
Can you create a separate job for the quality checks?

symbolist commented on requirements/circle.txt:
```diff
@@ -0,0 +1,7 @@
+-r test.txt
+
+codecov==2.0.15
+tox==3.1.3
```
This should be in `test.txt` so that the quality checks can be run locally too.

ArtemBernatskyy commented on .circleci/config.yml:
```diff
@@ -0,0 +1,69 @@
+          command: TOXENV=quality && tox
+      - run:
+          name: Generating code coverage
+          command: codecov
```
`codecov` only uploads results from `pytest-cov` to cloud in order to be visible to everybody (to include coverage report to `README.md` or etc).

It uses coverage 'artifacts' which were generated by `pytest-cov`. 
No config is needed for open source projects.
There is no point to run it locally because it will update coverage based on 'uncommited' code.

symbolist replied:
I am not sure how is codecov going to get the artifacts from a parallel job (which may not even have completed at this point). Or am I missing something?

We need local diff coverage so that when doing dev one can see if the tests have added the necessary coverage to the code that has been added. Can this do that? If not, we need to add diff-cover like in the blockstore repo I linked to.

symbolist replied:
Okay, I realize the later was not part of the scope of the task but it is related so lets get it in as well.

ArtemBernatskyy replied:
Yeah, @symbolist you are right, i've moved `codecov` to other task and added local `diff_cover`.

ArtemBernatskyy commented:
@symbolist `Circle CI` crashes if no tests are present, so i included empty test, or do i need to allow it to crash ?

symbolist commented:
@ArtemBernatskyy Its okay to include a test; it can be removed later.

symbolist commented on Makefile:
```diff
@@ -0,0 +1,33 @@
+	rm -fr *.egg-info
+
+diff_cover: test ## find diff lines that need test coverage
+	diff-cover coverage.xml
```
Also add this line to the end of test command so that whenever someone runs `make test` they get the diff coverage summary as well.

symbolist commented:
@ArtemBernatskyy 

1. In the Test Quality step of the quality job, I am seeing test results and not quality results?
2. In the test job, I see "Coverage.py warning: No data was collected. (no-data-collected)". Any idea why? Maybe try adding a test for `xblock_html/xblock_html.py` to see that the coverage is being tracked correctly?

ArtemBernatskyy commented:
@symbolist 
1. Fixed that it actually runs. But now it throws `InvocationError ` and i don't have any clue why it fails via `tox` and works via manually running command.
2. I've added import in order to simulate test and yes it actually calculates coverage.

symbolist commented:
@ArtemBernatskyy What is a code 20 error? Is the error because there are quality issues in the code? Can you please investigate this yourself and come up with a fix?

symbolist commented:
@ArtemBernatskyy Also all builds need to be green before merging so the quality issues need to be resolved as well.

ahmedaljazzar commented:
@ArtemBernatskyy We made a big merge in the master that caused conflicts in three files in your changes. If you need any clarifications during the rebase please don't hesitate to reach out.

symbolist commented on Makefile:
```diff
@@ -21,7 +21,7 @@ selfcheck: ## check that the Makefile is well-formed
 
 requirements:
-	pip install -q -r requirements.txt --exists-action w
+	pip install -q -r requirements/base.txt --exists-action w
```
I think we can make this `quality.txt` so that it installs everything needed for dev.

symbolist replied:
Also below in `quality` we need to change `pep8` with `pycodestyle`.
And actually run the tests in `test` + generate diff-cover. See https://github.com/open-craft/blockstore/blob/master/Makefile#L53

symbolist replied:
Also need to the add the commands to .PHONY.

symbolist commented on requirements/quality.txt:
```diff
@@ -0,0 +1,11 @@
+pylint==1.7.1
+caniusepython3==7.0.0
+edx_lint==0.5.5
+pep8==1.7.1
```
`pycodestyle` replaces it so it can be removed.

symbolist commented on .circleci/config.yml:
```diff
@@ -0,0 +1,70 @@
+          command: echo 'source /tmp/workspace/env/bin/activate' >> $BASH_ENV
+      - run:
+          name: Test Quality
+          command: export TOXENV=quality && tox
```
I am still seeing unexpected errors+warnings in the CircleCI build log.
